File: scripts/algorithms/M/Minimum Score After Removals on a Tree/Minimum Score After Removals on a Tree.py
# Runtime: 1682 ms (Top 100.0%) | Memory: 18.40 MB (Top 71.7%)

class Solution:
    def minimumScore(self, nums: List[int], edges: List[List[int]]) -> int:        
        n = len(nums)
        graph = [[] for _ in range(n)]
        for u, v in edges: 
            graph[u].append(v)
            graph[v].append(u)
            
        def fn(u, p): 
            nonlocal t
            score[u] = nums[u]
            tin[u] = (t := t+1) # time to enter
            for v in graph[u]: 
                if v != p: 
                    fn(v, u)
                    score[u] ^= score[v]
            tout[u] = t # time to exit 
        
        t = 0 
        score = [0]*n
        tin = [0]*n 
        tout = [0]*n 
        fn(0, -1)
        
        ans = inf 
        for u in range(1, n): 
            for v in range(u+1, n): 
                if tin[v] <= tin[u] and tout[v] >= tout[u]: # enter earlier & exit later == parent 
                    uu = score[u]
                    vv = score[v] ^ score[u]
                    xx = score[0] ^ score[v]
                elif tin[v] >= tin[u] and tout[v] <= tout[u]: 
                    uu = score[u] ^ score[v]
                    vv = score[v]
                    xx = score[0] ^ score[u]
                else: 
                    uu = score[u]
                    vv = score[v]
                    xx = score[0] ^ score[u] ^ score[v]
                ans = min(ans, max(uu, vv, xx) - min(uu, vv, xx))
        return ans  

File: scripts/algorithms/M/Minimum Score After Removals on a Tree/Minimum Score After Removals on a Tree.cpp
// Runtime: 167 ms (Top 87.54%) | Memory: 18 MB (Top 34.01%)
class Solution {
    vector<bool> visited;
    vector<vector<int>> pc, adj; // pc is the parent child edge in our dfs
    vector<vector<bool>> childs; // 2D array to store that i is a parent of j
    vector<int> child_xor, nums, par; // child_xor to store result of xors of a child node and par is a gloable array to track the parents of a node in dfs

    int dfs(int i) {
        int ans = nums[i];
        visited[i] = true;

        for (int p: par) childs[p][i] = true; // Defining this node as the child of all its parents

        par.push_back(i);

        for (int child: adj[i])
            if (!visited[child]) {
                pc.push_back({i, child});
                ans ^= dfs(child); // Recurcively calculating xors
            }

        par.pop_back();

        return child_xor[i] = ans;
    }
public:
    int minimumScore(vector<int>& nums, vector<vector<int>>& edges) {
        // Initialising gloable variables
        int n = nums.size(), ans = INT_MAX;
        visited = vector<bool>(n);
        pc = vector<vector<int>>();
        adj = vector<vector<int>>(n);
        child_xor = vector<int>(n, 0);
        childs = vector<vector<bool>>(n, vector<bool>(n, false));
        this->nums = nums;
        par = vector<int>();

        // Creating an adjacency matrix
        for (vector<int> &edge: edges) adj[edge[0]].push_back(edge[1]), adj[edge[1]].push_back(edge[0]);

        dfs(0);

        for (int i = 0; i < pc.size(); i++)
            for (int j = i + 1; j < pc.size(); j++) { // removing an edge i and j
                int a = pc[i][1], b = pc[j][1]; // node that will come below when you delete an edge i and j
                int xa = child_xor[a], xb = child_xor[b], xc = child_xor[0];

                if (childs[a][b])
                    xc ^= xa, xa ^= xb;
                else
                    xc ^= xa, xc ^= xb;

                ans = min(max(xa, max(xb, xc)) - min(xa, min(xb, xc)), ans);
            }

        return ans;
    }
};